content/blog/2022-04-04-develop-and-test-data-analytics-pipelines-on-localstack/index.md

```diff
@@ -0,0 +1,236 @@
+---
+title: Develop and Test Real-time Data Pipelines with LocalStack
+description: "LocalStack makes it easy to develop and test real-time data pipelines that are built with AWS services. Learn how to deploy Cloudwatch, Kinesis, Lambda, and external services using CDK on LocalStack."
+lead: "LocalStack makes it easy to develop and test real-time data pipelines that are built with AWS services. Learn how to deploy Cloudwatch, Kinesis, Lambda, and external services using CDK on LocalStack."
+date: 2022-04-04
+lastmod: 2022-04-04
+draft: false
+images: []
+contributors: ["Sam Watson"]
+tags: ["tutorial"]
+---
+
+{{< img src="teaser.png" >}}
+
+At LocalStack, we rely on AWS Lambda as a key part of our serverless infrastructure toolkit. As with any critical service, we want to extract analytics event data from running Lambda functions and aggregate it in our data warehouse so that we can track user stories and squash bugs. But this presents a challenge: how can we emit detailed analytics data while keeping Lambda code simple and performant? To solve this, we developed a serverless streaming data pipeline using CloudWatch Logs and Kinesis that allows us to decouple decouple analytics from application logic.
+
+
+{{< img src="pipeline.png" >}}
+
+
+In this pipeline, a Lambda function writes analytics event data as structured JSON payloads to CloudWatch Logs. From there, we use a [subscription filter](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/Subscriptions.html) to extract the events from the function's log group and deliver them directly to a Kinesis stream. Finally, a separate Lambda function consumes events off the stream in batches and delivers them to our Tinybird data warehouse via their API.
```

Comment:
It would be recommended to add a note about Tinybird, and what it does alongside its capabilities.

Reply:
Could you please clarify a little bit what you have in mind here? In my mind saying that Tinybird is a "data warehouse" is sufficient to communicate what it does. I was trying to avoid getting too into the weeds on Tinybird in this article because I want the main focus to be on the pipeline, and the only Tinybird-specific feature we're making use of here is their API.

Reply:
@thrau has fixed it!

Reply:
makes sense @SwatsonCodes , i nevertheless added a brief sentence to give some context about tinybird. the folks at TB will also be happy about it :-)